data/blog/libtailscale.mdx

```diff
@@ -0,0 +1,198 @@
+---
+title: "libtailscale: how it's made"
+date: '2023-03-02'
+tags:
+  - 'tsnet'
+  - 'libtailscale'
+  - 'oss'
+authors: ['crawshaw']
+summary: 'libtailscale: a Tailscale C library'
+---
+
+We just released an experimental libtailscale, the Tailscale C library.
+It is built on the Go [tsnet](https://pkg.go.dev/tailscale.com/tsnet)
+package which uses a userland network stack to implement TCP/IP
+directly inside your process.
+
+```
+tailscale ts = tailscale_new();
+if (tailscale_set_ephemeral(ts, 1)) {
+	return err(ts);
+}
+if (tailscale_up(ts)) {
+	return err(ts);
+}
+tailscale_listener ln;
+if (tailscale_listen(ts, "tcp", ":1999", &ln)) {
+	return err(ts);
+}
+while (1) {
+	tailscale_conn conn;
+	if (tailscale_accept(ln, &conn)) {
+		return err(ts);
+	}
+	char buf[2048];
+	while ((ret = read(conn, buf, sizeof(buf))) > 0) {
+		write(1, buf, ret);
+	}
+	close(conn);
+}
+```
+
+For those who have spent some time in our source code, you’ll know
+Tailscale writes almost everything in Go.
+So how did we end up with a C library?
+It’s not widely used, but the standard Go toolchain includes all the
+machinery necessary for producing a C library in the form of
+[build modes](https://pkg.go.dev/cmd/go#hdr-Build_modes).
+
+## Go Build Modes
+
+The two key build modes are `c-archive` and `c-shared`.
+The first creates an object archive that can be linked by a standard
+C toolchain into a binary.
+The second creates a shared object (on linux, the .so file) that can
+be dynamically linked into a binary.
+They are surprisingly easily to use:
+
+    go build -buildmode=c-archive pkg
+
+This produces an object archive, and a generated header for the C
+symbols Go exports.
+
+As part of compilation, cgo can “export” your Go functions to C using
```

Comment:
Curious to know if the use of curly quotes in “export” render properly.  Might be fine, just something to check.

Reply:
switched to ordinary quotes (I wrote this in Apple Notes and copied it in)